src/dtls.c

```diff
@@ -290,60 +354,459 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+                                            (word16)tlsx.size, &parsedExts);
+            if (ret != 0)
+                goto dtls13_cleanup;
+#if defined(HAVE_SESSION_TICKET) || !defined(NO_PSK)
+            haveKS = 1;
+#endif
+        }
+
+#if defined(HAVE_SESSION_TICKET) || !defined(NO_PSK)
+        /* Pre-shared key */
+        ret = TlsxFindByType(&tlsx, TLSX_PRE_SHARED_KEY, ch->extension);
+        if (ret != 0)
+            goto dtls13_cleanup;
+        if (tlsx.size != 0) {
+            /* Let's just assume that the binders are correct here. We will
+             * actually verify this in the stateful part of the processing
+             * and if they don't match we will error out there anyway. */
+            byte modes;
+
+#ifndef NO_PSK
+            /* When we didn't find a valid ticket ask the user for the
+             * ciphersuite matching this identity */
+            if (!pskInfo->isValid) {
+                if (TLSX_PreSharedKey_Parse_ClientHello(&parsedExts,
+                        tlsx.elements, tlsx.size, ssl->heap) == 0)
+                    FindPskSuiteFromExt(ssl, parsedExts, pskInfo, &suites);
+                /* Revert to full handshake if PSK parsing failed */
+            }
+#endif
+
+            if (pskInfo->isValid) {
+                ret = TlsxFindByType(&tlsx, TLSX_PSK_KEY_EXCHANGE_MODES,
+                                     ch->extension);
+                if (ret != 0)
+                    goto dtls13_cleanup;
+                if (tlsx.size == 0)
+                    ERROR_OUT(MISSING_HANDSHAKE_DATA, dtls13_cleanup);
+                ret = TLSX_PskKeyModes_Parse_Modes(tlsx.elements, tlsx.size,
+                                                  client_hello, &modes);
+                if (ret != 0)
+                    goto dtls13_cleanup;
+                if ((modes & (1 << PSK_DHE_KE)) &&
+                        !ssl->options.noPskDheKe) {
+                    if (!haveKS)
+                        ERROR_OUT(MISSING_HANDSHAKE_DATA, dtls13_cleanup);
+                    doKE = 1;
+                }
+                else if ((modes & (1 << PSK_KE)) == 0) {
+                        ERROR_OUT(PSK_KEY_ERROR, dtls13_cleanup);
+                }
+                usePSK = 1;
+            }
+        }
+#endif
+
+#if defined(HAVE_SESSION_TICKET) || !defined(NO_PSK)
+        if (usePSK && pskInfo->isValid) {
+            cs.cipherSuite0 = pskInfo->cipherSuite0;
+            cs.cipherSuite = pskInfo->cipherSuite;
+
+            if (doKE) {
+                byte searched = 0;
+                ret = TLSX_KeyShare_Choose(ssl, parsedExts, &cs.clientKSE,
+                        &searched);
+                if (ret != 0)
+                    goto dtls13_cleanup;
+                if (cs.clientKSE == NULL && searched)
+                    cs.doHelloRetry = 1;
+            }
+        }
+        else
+#endif
+        {
+            ret = MatchSuite_ex(ssl, &suites, &cs, parsedExts);
```

Comment:
In case the client isn't using PSK based handshake. Do we check that he provides a good keyshare? I don't see `TLSX_KeyShare_Choose` invoked in the non-PSK code path

Reply:
`TLSX_KeyShare_Choose` is invoked deep in `MatchSuite_ex`.